File: LeetCodePractice/src/backtracking/PermutationsII.java
package backtracking;

/*Given a collection of numbers that might contain duplicates, return all possible unique permutations.

Example:

Input: [1,1,2]
Output:
[
  [1,1,2],
  [1,2,1],
  [2,1,1]
]*/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class PermutationsII {
    /**
     * @param nums: A list of integers.
     * @return: A list of unique permutations.
     */
    public List<List<Integer>> permuteUnique(int[] nums) {
        // Write your code here
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null) {
            return res;
        }
        if (nums.length == 0) {
            res.add(new ArrayList<>());
            return res;
        }
        Arrays.sort(nums);
        int len = nums.length;
        boolean[] visited = new boolean[len];
        List<Integer> temp = new ArrayList<>();
        dfs(res, temp, nums, visited);
        return res;
    }

    private void dfs(List<List<Integer>> res, List<Integer> temp, int[] nums, boolean[] visited) {
        if (temp.size() == nums.length) {
            res.add(new ArrayList<>(temp));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!visited[i]) {
                if (i > 0 && nums[i] == nums[i - 1] && visited[i - 1]) {
                    continue;
                }
                visited[i] = true;
                temp.add(nums[i]);
                dfs(res, temp, nums, visited);
                visited[i] = false;
                temp.remove(temp.size() - 1);
            }
        }
    }
}

File: LeetCodePython/PermutationsII.py
"""
Given a collection of numbers, nums, that might contain duplicates, return all possible unique permutations in any order.

Example 1:
Input: nums = [1,1,2]
Output:
[[1,1,2],
 [1,2,1],
 [2,1,1]]

Example 2:
Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]


Constraints:
1 <= nums.length <= 8
-10 <= nums[i] <= 10
"""
from typing import List


class PermutationsII:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        visited = [False] * len(nums)

        def dfs(ls):
            if len(ls) == len(nums):
                res.append(ls.copy())
                return
            for i, n in enumerate(nums):
                if not visited[i]:
                    if i > 0 and nums[i] == nums[i - 1] and not visited[i - 1]:
                        continue

                    visited[i] = True
                    ls.append(n)
                    dfs(ls)
                    ls.pop()
                    visited[i] = False

        dfs([])
        return res
